src/utils/DatabaseFilter.cpp

```diff
@@ -0,0 +1,167 @@
+
+class TOPPDatabaseFilter :
+  public TOPPBase
+{
+public:
+  TOPPDatabaseFilter() :
+    TOPPBase("DatabaseFilter", "Tool to create a reduced database.", false)
+  {
+
+  }
+
+protected:
+  void registerOptionsAndFlags_()
+  {
+    registerInputFile_("in", "<file>", "","Input FASTA file, containing a database.");
+    setValidFormats_("in", ListUtils::create<String>("fasta"));
+    registerInputFile_("accession", "<file>", "", "Input IdXML file, containing the identfied peptides.", true);
+    setValidFormats_("accession", ListUtils::create<String>("idXML"));
+    registerStringOption_("method", "<type>", "whitelist", "Switch between white/blacklisting", false);
+    setValidStrings_("method", ListUtils::create<String>("whitelist,blacklist"));
+    registerOutputFile_("out", "<file>", "", "Output FASTA file where the reduced database will be written to.");
+    setValidFormats_("out", ListUtils::create<String>("fasta"));
+  }
+
+  ExitCodes main_(int, const char **)
+  {
+    //-------------------------------------------------------------
+    // parsing parameters
+    //-------------------------------------------------------------
+    String in(getStringOption_("in"));
+    String ids(getStringOption_("accession"));
+    String method(getStringOption_("method"));
+    bool whitelist = (method == "whitelist");
+    String out(getStringOption_("out"));
+
+    //-------------------------------------------------------------
+    // reading input
+    //-------------------------------------------------------------
+    vector<FASTAFile::FASTAEntry> db;
+    FASTAFile ().load(in, db);
+
+    vector<ProteinIdentification> protein_identifications;
+    vector<PeptideIdentification> peptide_identifications;
+    IdXMLFile().load(ids, protein_identifications, peptide_identifications);
+
+    LOG_INFO << "Identifications: " << ids.size() << endl;
+
+    //-------------------------------------------------------------
+    // calculations
+    //-------------------------------------------------------------
+
+    vector<FASTAFile::FASTAEntry> db_new;
+
+    set<String> id_accessions;
+    for (Size i = 0; i != peptide_identifications.size(); ++i)
+    {
+      const PeptideIdentification& id = peptide_identifications[i];
+      const vector<PeptideHit>& hits = id.getHits();
+      for (Size k = 0; k != hits.size(); ++k)
+      {
+        const vector<PeptideEvidence>& evidences = hits[k].getPeptideEvidences();
+        for (Size m = 0; m != hits.size(); ++m)
+        {
+          const String& id_accession = evidences[m].getProteinAccession();
+          id_accessions.insert(id_accession);
+        }
+      }
+    }
+
+    LOG_INFO << "Protein accessions: " << id_accessions.size() << endl;
+
+    for (Size i = 0; i != db.size() ; ++i)
+    {
+      const String& fasta_accession = db[i].identifier;
+      const bool found = id_accessions.find(fasta_accession) != id_accessions.end();
+      if ( (found && whitelist) || (!found && !whitelist) )
```

Comment:
maybe add a comment here: "either found in the whitelist or not found in the blacklist"